docs/book/src/developer/providers/ipam.md

```diff
@@ -0,0 +1,81 @@
+
+## Behaviour
+
+IPAM providers must handle any IPAddressClaim resources that reference IP address pools that are managed by the provider and create an IPAddress resource for it. IPAddressClaims are usually created by infrastructure providers.
+
+### IPAM Provider
+
+An IPAM provider must watch for new, updated and deleted IPAddressClaims that reference an IP address pool that is manged by the provider in their `spec.poolRef` field.
+
+#### Normal IPAddressClaim
+
+1. If the IPAddressClaim does not reference a pool managed by the provider in it's `spec.poolRef`, abort the reconciliation.
+2. If the related Cluster is paused, abort reconciliation
+   1. The related Cluster is referenced using the `spec.clusterName` field or a `cluster.x-k8s.io/cluster-name` label (the latter is deprecated).
+   2. If the field is empty and the annotation is not present, reconciliation can continue.
+   3. If the referenced cluster is not found, abort reconciliation.
+   4. If the referenced cluster has `spec.paused` set or a `cluster.x-k8s.io/paused` annotation, skip reconciliation
+3. Add any required provider-specific Finalziers (you probably need one)
+4. Allocate an IP address for the claim
+5. Create an IPAddress object
+   1. It should have the same name as the claim.
+   2. It must have a owner reference with `controller: true` and `blockOwnerDeletion: true` to the Claim
+   3. It must have a owner reference with `controller: false` and `blockOwnerDeletion: true` to the referenced Pool
+   4. It should have a Finalizer that prevents accidental deletion, e.g. `ipam.cluster.x-k8s.io/protect-address`.
+6. Set the `status.addressRef` to the created IPAddress
+
+#### Deleted IPAddressClaim
+
+1. If the related Cluster is paused, abort reconciliation (see 2. above)
+2. Deallocate the IP address
+3. Delete the IPAddress object
+   1. Remove any Finalizers that were set to prevent deletion first
+4. Remove the Finalizer from the claim
+
+#### Clusterctl Move
+
+In order for Pools to be moved alongside clusters, they need to have a `cluster.x-k8s.io/cluster-name` label.
+
+### Infrastructure Provider
+
+In order to consume IP addresses from an IP address pool, an IPAddressClaim resource needs to be created, which will then be fulfilled with an IPAddress resource. Since the IPAddressClaim needs to reference an IP pool, you'll need to add a property to your infrastructure Machine that allows to specify the pool.
+
+<aside class="note">
+
+Note that the IPAM contract is single-stack. If you need both v4 and v6 addresses, two pools and two IPAddressClaims are necessary.
+
+</aside>
+
+1. Create an IPAddressClaim
+   1. The `spec.poolRef` must reference the pool you want to use
+   2. It should have an owner reference to the infrastructure Machine it is created for (required to support `clusterctl move`)
```

Comment:
What about controller / blockOwnerDeletion here?

Reply:
Good point. We're currently investigating an issue when deleting metal3 based clusters where claims are deadlocked due to the cluster vanishing before they are cleaned up. The paused check prevents the claim from being released if the cluster can't be found. We'll either have to make sure that the cluster doesn't get deleted, or ignore the paused check when releasing addresses.

Reply:
I propose to merge this as is, and we'll update it if we change how this works in the in-cluster ipam implementation.